research/src/pages/file.proposal.mdx

```diff
@@ -0,0 +1,165 @@
+
+The `<input type=file>` control is widely used to provide applications with the
+capability of reading local files on form submission, or uploading files to a
+server.
+
+### Features <a href="#features" id="features"></a>
+
+The main feature is selecting local files, potentially receiving hints from the
+website about which file kinds are allowed, for example.
+
+### Prior Art/Examples <a href="#prior-art" id="prior-art"></a>
+
+#### Browser implementations
+
+Browsers show an interoperable rendering for file inputs, consisting of a
+button and a label.
+
+- WebKit and Blink use an `<input type=button>`, and a magical label that
+  doesn't show up in the DOM.
+
+- Gecko uses an actual `<button tabindex=-1>` and an HTML `<label>` element.
+
+Those three implementations use some form of middle-cropping when the selected
+file name is too large.
+
+#### Custom inputs on the web
+
+Custom inputs on the web are generally achieved by hiding the original input.
+Some approach it by hiding it in a way that the user can still hit (like
+`opacity: 0`), others use JS to forward the click from some other UI element to
+the underlying file input.
+
+- [Bootstrap](https://getbootstrap.com/docs/4.0/components/forms/#file-browser)
+  uses the former approach, by generating a "Browse" button with generated
+  content.
+
+- [Carbon Design System](https://www.carbondesignsystem.com/components/file-uploader/code/)
+  uses the later.
+
+---
+
+### API <a href="#api" id="api"></a>
+
+See [the Resources section](#resources) for the specification and the MDN reference.
+
+#### Properties and Attributes <a href="#properties-attributes" id="properties-attributes"></a>
+
+See [the Resources section](#resources) for the specification and the MDN reference.
+
+#### Events <a href="#events" id="events"></a>
+
+See [the Resources section](#resources) for the specification and the MDN reference.
+
+### Appearance <a href="#appearance" id="appearance"></a>
+
+![File Chooser Example](../images/file.png)
+
+### Anatomy <a href="#anatomy" id="anatomy"></a>
+
+TODO
+
+#### Diagram <a href="#diagram" id="diagram"></a>
+
+TODO
+
+#### DOM Structure <a href="#dom-structure" id="dom-structure"></a>
+
+```html
+<button tabindex="-1">Browse...</button>
+<!-- Or appropriately-localized string -->
+<label><!-- implementation and state-dependent --></label>
+```
+
+It is important that the button is not tab-navigable, as the input is focused
+atomically, see below.
+
+#### CSS Parts <a href="#css-parts" id="css-parts"></a>
+
+| Part Name             | Description                                                                                                                     |
+| --------------------- | ------------------------------------------------------------------------------------------------------------------------------- |
+| `file-chooser-button` | Enables styling the `&lt;button>` or `&lt;input type=button>` part without requiring it to be completely replaced via its slot. |
```

Comment:
Looking through the different parts across the different component libraries:

* File Picker
* File Selector

I think both of these are better than chooser but I do think all of them should be appended with button

Reply:
I agree with the `-button` suffix as the selector's purpose essentially dictates it. However, for consistency's sake, the verb really should be a choice between "upload" and "select"/"selector". The majority of standards documents for HTML and JavaScript refer to such controls as file upload, and the action performed as selection, and have done so for many years. What benefit is there to CSS authors in deviating from those conventions?

There is merit to each:

* the button triggers a file selection system but operating it doesn't guarantee a file upload attempt by the user; and
* emphasising upload as the action is important because successful selection of any file immediately makes the file's name and contents available to running script, unlike the olden days when it was only a state of intent to upload upon form submission.

Reply:
@Roy-Orbison that seems like a valid argument for selector then. One aspect that Open UI is trying to ensure is that it uses nomenclature used by web developers building controls/components today as well as industry standards or other operating system paradigms, etc. So based on this I think `file-selector-button` seems like a good fit. @emilio what do you think?